python/cuda_parallel/cuda/parallel/experimental/algorithms/reduce.py

```diff
@@ -46,6 +46,30 @@ def _dtype_validation(dt1, dt2):
         raise TypeError(f"dtype mismatch: __init__={dt1}, __call__={dt2}")
 
 
+def _validate_and_get_stream(stream) -> Optional[int]:
+    # null stream is allowed
+    if stream is None:
+        return None
+
+    if not hasattr(stream, "__cuda_stream__"):
+        raise TypeError(
+            f"stream argument {stream} does not implement the '__cuda_stream__' protocol"
+        )
+
+    stream_property = stream.__cuda_stream__
+    if (
+        isinstance(stream_property, tuple)
+        and len(stream_property) == 2
+        and all(isinstance(i, int) for i in stream_property)
+    ):
+        version, handle = stream_property
+        return handle
+
+    raise TypeError(
+        f"__cuda_stream__ property of '{stream}' must return a 'Tuple[int, int]'; got {stream_property} instead"
+    )
```

Comment:
nit: Personally, I'd be a bit more lax here and really only ensure that `handle` is an int. If someone returns a `list` rather than a `tuple`, that's probably not relevant to us:

```suggestion
    _, handle = stream_property  # (version, handle)
    if not isinstance(handle, int):
        raise TypeError(...)
    return handle
```

Reply:
We should also check if version is 0, because we could change it to say 3-tuple in a future version of the protocol.